### ISSUE TITLE
v2.0: dies on out-of-memory on single file PUT

### ISSUE THREAD

**Issue by zarnovican** [bloomreach/s4cmd#65]
Hi,

while trying this tool (for the first time) to `dsync` 80GB of data (in 50 files) from local to S3, it consumed all the memory and was killed by OOM killer.

```
s4cmd dsync -r . s3://xxx/normalize_exports/
```

Then, I tried to PUT individual files one-by-one and it also run out of memory :disappointed: 

```
$ ls -lh 5s/2015-07.jsonl.gz
-rw-rw-r-- 1 ubuntu ubuntu 808M May 25 06:39 5s/2015-07.jsonl.gz
$ s4cmd put 5s/2015-07.jsonl.gz s3://xxx/normalize_exports/5s/2015-07.jsonl.gz
```

Versions:

```
$ pip list
awscli (1.10.33)
boto3 (1.3.1)
botocore (1.4.23)
colorama (0.3.3)
docutils (0.12)
futures (3.0.5)
jmespath (0.9.0)
pip (8.1.2)
pyasn1 (0.1.9)
python-dateutil (2.5.3)
pytz (2016.4)
rsa (3.4.2)
s3transfer (0.0.1)
s4cmd (2.0.1)
setuptools (2.2)
six (1.10.0)
```

**Comment by TTimo** [comment]
I haven't tried on a set of smaller files, but I can confirm s4cmd dies out of memory on large file uploads - this particular one it failed on was a 450GB backup archive.

**Comment by TTimo** [comment]
Same error happens on the same data set, but split into roughly 100 5GB files. So it's not the size of the individual file.

tl;dr - s4cmd is not working for large transfers - this drastically reduces it's usefulness

**Comment by TTimo** [comment]
Sorry I've moved to using other tools for s3 transfers since. I don't remember the particulars of reproducing this.

**Comment by zarnovican** [comment]
> Sorry I've moved to using other tools for s3 transfers since. I don't remember the particulars of reproducing this.

Same here. Different tools and company as well. I'm ok to close this ticket.

**Comment by navinpai** [comment]
Alright guys... Thanks for the update, and hope you guys find s4cmd useful sometime soon!

Will go ahead and close this issue. :+1: 

Just out of curiosity: What tools do you use for s3 transfers now?

**Comment by TTimo** [comment]
Just the aws cli these days.

**Comment by pvgoran** [comment]
Uploading file of 6.3G, with 24 threads, takes 6000M of RSS, as per `htop`.

The machine has 16GB of RAM.

**Comment by pvgoran** [comment]
Limiting concurrency to 4 threads lowers max consumption to a bit less than 2GB. This is still too much for a simple transfer task, though. And consuming 6GB by default (without options) makes the tool too dangerous to use.

My environment: NixOS 18.09, python3.
